nmigen_boards/orangecrab_r0_2.py

```diff
@@ -0,0 +1,127 @@
+        # Used to reload FPGA configuration.
+        # Can enter USB bootloader by assigning button 0 to program.
+        Resource("program", 0, PinsN("V17", dir="o"), Attrs(IO_TYPE="LVCMOS33")),
+
+        RGBLEDResource(0, r="K4", g="M3", b="J3", invert=True,
+                      attrs=Attrs(IO_TYPE="LVCMOS33")),
+
+        *ButtonResources(pins={0: "J17" }, invert=True,
+                         attrs=Attrs(IO_TYPE="SSTL135_I")),
+
+        *SPIFlashResources(0,
+                           cs="U17", clk="U16", miso="T18", mosi="U18", wp="R18", hold="N18",
+                           attrs=Attrs(IO_STANDARD="LVCMOS33")
+        ),
+
+        Resource("ddr3", 0,
+            Subsignal("rst",     PinsN("L18", dir="o")),
+            Subsignal("clk",     DiffPairs("J18", "K18", dir="o"), Attrs(IO_TYPE="SSTL135D_I")),
+            Subsignal("clk_en",  Pins("D18", dir="o")),
+            Subsignal("cs",      PinsN("A12", dir="o")),
+            Subsignal("we",      PinsN("B12", dir="o")),
+            Subsignal("ras",     PinsN("C12", dir="o")),
+            Subsignal("cas",     PinsN("D13", dir="o")),
+            Subsignal("a",       Pins("C4 D2 D3 A3 A4 D4 C3 B2 B1 D1 A7 C2 B6 C1 A2 C7", dir="o")),
+            Subsignal("ba",      Pins("P5 N3 M3", dir="o")),
+            Subsignal("dqs",     DiffPairs("G18 H17", "B15 A16", dir="io"),
+                      Attrs(IO_TYPE="SSTL135D_I", TERMINATION="OFF",
+                      DIFFRESISTOR="100")),
+            Subsignal("dq",      Pins("C17 D15 B17 C16 A15 B13 A17 A13 F17 F16 G15 F15 J16 C18"
+                                      "H16 F18", dir="io"), Attrs(TERMINATION="75")),
+            Subsignal("dm",      Pins("G16 D16", dir="o")),
+            Subsignal("odt",     Pins("C13", dir="o")),
+            Attrs(IO_TYPE="SSTL135_I", SLEWRATE="FAST")
+        ),
+
+        Resource("ddr3_pseudo_power", 0,
+            # pseudo power pins, leave these at their default value
+            Subsignal("vcc_virtual", PinsN("K16 D17 K15 K17 B18 C6", dir="o")),
+            Subsignal("gnd_virtual", Pins("L15 L16", dir="o")),
+            Attrs(IO_TYPE="SSTL135_II", SLEWRATE="FAST")
+        ),
+
+        Resource("adc", 0,
+            Subsignal("ctrl",     Pins("G1 F1", dir="o")),
+            Subsignal("mux",      Pins("F4 F3 F2 H1", dir="o")),
+            Subsignal("sense_hi", Pins("H3", dir="i")),
+            Subsignal("sense_lo", Pins("G3", dir="i")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+
+        *SDCardResources(0, dat0="J1", dat1="K3", dat2="L3", dat3="M1", clk="K1", cmd="K2", cd="L1",
+            attrs=Attrs(IO_TYPE="LVCMOS33", SLEWRATE="FAST")
+        ),
+
+        Resource("usb", 0,
+            Subsignal("d_p",    Pins("N1", dir="io")),
+            Subsignal("d_m",    Pins("M2", dir="io")),
+            Subsignal("pullup", Pins("N2", dir="o")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+
+        Resource("spi", 0,
```

Comment:
This should use `SPIResource` I think

Reply:
And definitely not have all directions set to `io` (which is the exact thing that `SPIResources` exists to avoid)

Reply:
SPIResources requires a cs pin, which I don't have.

Reply:
Ah I see. That actually makes it not especially useful for SPI in general, since lots of resources use the CS pin for framing. In fact I would hesitate to call it "SPI" at all?

Reply:
It's just GPIOs labeled "spi", I think you're supposed to combine it with any other numbered pin as a CS. I do agree that CS is important and I wouldn't make CS optional for SPIResources.

Reply:
I think the reason for this is it copies the Adafruit Feather pin names. The AVR based Feathers have hardware SPI on those pins.

Reply:
Yeah, the pins are in the location of hardware SPI found on other boards, they are shared with standard GPIOs.
If someone wanted to use a SPI module that they would need to add another GPIO as the CS pin.

Reply:
Ah I see! Now that I look closer at how the Feather looks like, I think I understand better what would work here. Let's do this:
  * scrap the Resources for SPI and I2C;
  * make one huge Connector for the entire set of PTH, with some pins named numerically (0, 1, 2...), and some symbolically (SCK, MO, MI, SDA, SCL).

Connectors treat pin names as black boxes basically, there are no restrictions on them as long as they're unique, so you can just put whatever's on the silkscreen there and it will work.

Reply:
I did it all with strings. If you mean mixing data types, nmigen's Connector class has a limitation that a provided dict must be keyed with strings.

Reply:
Sorry, I actually meant that nMigen treats all connector pins as opaque strings, even if you specify them as a single sequence (for connectors with pins numbered from 1 to n). So, since it doesn't try to give the strings any further meaning, you can freely mix numbers and letters.